d-sandbox
<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 400px">
</div>

# Spark SQL
1. Run a SQL query
1. Create a DataFrame
1. Write same query using DataFrame transformations
1. Trigger computation with DataFrame actions
1. Convert between DataFrames and SQL

##### Methods
- SparkSession (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=sparksession#pyspark.sql.SparkSession" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">Scala</a>): `sql`, `table`
- DataFrame (<a href="https://spark.apache.org/docs/latest/api/python/pyspark.sql.html?highlight=dataframe#pyspark.sql.DataFrame" target="_blank">Python</a>/<a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">Scala</a>):
  - Transformations:  `select`, `where`, `orderBy`
  - Actions: `show`, `count`, `take`
  - Other methods: `printSchema`, `schema`, `createOrReplaceTempView`

In [0]:
%run ./Includes/Classroom-Setup-SQL

Datasets mounted and student environment set up

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Run a SQL query
Use `SparkSession` to run SQL

In [0]:
budgetDF = spark.sql("""
SELECT name, price
FROM products
WHERE price < 200
ORDER BY price
""")

View results in the returned DataFrame

In [0]:
budgetDF.show()

+--------------------+-----+
 name|price|
+--------------------+-----+
Standard Foam Pillow| 59.0|
 King Foam Pillow| 79.0|
Standard Down Pillow|119.0|
 King Down Pillow|159.0|
+--------------------+-----+

In [0]:
display(budgetDF)

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Create a DataFrame
Use `SparkSession` to create a DataFrame from a table

In [0]:
productsDF = spark.table("products")
display(productsDF)

item_id name price M_PREM_Q Premium Queen Mattress 1795.0 M_STAN_F Standard Full Mattress 945.0 M_PREM_F Premium Full Mattress 1695.0 M_PREM_T Premium Twin Mattress 1095.0 M_PREM_K Premium King Mattress 1995.0 P_DOWN_S Standard Down Pillow 119.0 M_STAN_Q Standard Queen Mattress 1045.0 M_STAN_K Standard King Mattress 1195.0 M_STAN_T Standard Twin Mattress 595.0 P_FOAM_S Standard Foam Pillow 59.0 P_FOAM_K King Foam Pillow 79.0 P_DOWN_K King Down Pillow 159.0

Access schema of DataFrame

In [0]:
productsDF.printSchema()

root
-- item_id: string (nullable = true)
-- name: string (nullable = true)
-- price: double (nullable = true)

In [0]:
productsDF.schema

Out[10]: StructType(List(StructField(item_id,StringType,true),StructField(name,StringType,true),StructField(price,DoubleType,true)))

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Write same query with DataFrame transformations

In [0]:
budgetDF = (productsDF
  .select("name", "price")
  .where("price < 200")
  .orderBy("price")
)

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Trigger computation with DataFrame actions

In [0]:
budgetDF.count()

Out[12]: 4

In [0]:
budgetDF.take(2)

Out[13]: [Row(name='Standard Foam Pillow', price=59.0),
 Row(name='King Foam Pillow', price=79.0)]

### ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Convert between DataFrames and SQL

In [0]:
budgetDF.createOrReplaceTempView("budget")

In [0]:
display(spark.sql("SELECT * FROM budget"))

name price Standard Foam Pillow 59.0 King Foam Pillow 79.0 Standard Down Pillow 119.0 King Down Pillow 159.0

## ![Spark Logo Tiny](https://files.training.databricks.com/images/105/logo_spark_tiny.png) Lab

1. Create a DataFrame from the `Event` table
1. Display DataFrame and inspect schema
1. Apply transformations to filter and sort `macOS` events
1. Count results and take first 5 rows
1. Create the same DataFrame using SQL query

##### Methods
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/SparkSession.html" target="_blank">SparkSession</a>: `sql`, `table`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> transformations: `select`, `where`, `orderBy`
- <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> actions: `select`, `count`, `take`
- Other <a href="http://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html" target="_blank">DataFrame</a> methods: `printSchema`, `schema`, `createOrReplaceTempView`

### 1. Create a DataFrame from the `events` table
- Use SparkSession to create a DataFrame from the `events` table

In [0]:
# TODO
eventDF = spark.table("events")

### 2. Display DataFrame and inspect schema
- Use methods above to inspect DataFrame contents and schema

In [0]:
display(eventDF.limit(5))

device ecommerce event_name event_previous_timestamp event_timestamp geo items traffic_source user_first_touch_timestamp user_id macOS List(null, null, null) warranty 1593878899217692 1593878946592107 List(Montrose, MI) List() google 1593878899217692 UA000000107379500 Windows List(null, null, null) press 1593876662175340 1593877011756535 List(Northampton, MA) List() google 1593876662175340 UA000000107359357 macOS List(null, null, null) add_item 1593878792892652 1593878815459100 List(Salinas, CA) List(List(null, M_STAN_T, Standard Twin Mattress, 595.0, 595.0, 1)) youtube 1593878455472030 UA000000107375547 iOS List(null, null, null) mattresses 1593878178791663 1593878809276923 List(Everett, MA) List() facebook 1593877903116176 UA000000107370581 Windows List(null, null, null) mattresses null 1593878628143633 List(Cottage Grove, MN) List() google 1593878628143633 UA000000107377108

In [0]:
eventDF.printSchema()

root
-- device: string (nullable = true)
-- ecommerce: struct (nullable = true)
 |-- purchase_revenue_in_usd: double (nullable = true)
 |-- total_item_quantity: long (nullable = true)
 |-- unique_items: long (nullable = true)
-- event_name: string (nullable = true)
-- event_previous_timestamp: long (nullable = true)
-- event_timestamp: long (nullable = true)
-- geo: struct (nullable = true)
 |-- city: string (nullable = true)
 |-- state: string (nullable = true)
-- items: array (nullable = true)
 |-- element: struct (containsNull = true)
 | |-- coupon: string (nullable = true)
 | |-- item_id: string (nullable = true)
 | |-- item_name: string (nullable = true)
 | |-- item_revenue_in_usd: double (nullable = true)
 | |-- price_in_usd: double (nullable = true)
 | |-- quantity: long (nullable = true)
-- traffic_source: string (nullable = true)
-- user_first_touch_timestamp: long (nullable = true)
-- user_id: string (nullable = true)

-sandbox
### 3. Apply transformations to filter and sort `macOS` events
- Filter for rows where `device` is `macOS`
- Sort rows by `event_timestamp`

<img alt="Hint" title="Hint" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-light-bulb.svg"/>&nbsp;**Hint:** Use single and double quotes in your filter SQL expression

In [0]:
from pyspark.sql.functions import *
macDF = (eventDF
         .filter(col("device") == "macOS")
         .orderBy("event_timestamp")
        )

### 4. Count results and take first 5 rows
- Use DataFrame actions to count and take rows

In [0]:
# TODO
numRows = macDF.count()
rows = macDF.take(5)

-sandbox
##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work

In [0]:
from pyspark.sql import Row

assert(numRows == 1938215)
assert(len(rows) == 5)
assert(type(rows[0]) == Row)

### 5. Create the same DataFrame using SQL query
- Use SparkSession to run a sql query on the `events` table
- Use SQL commands above to write the same filter and sort query used earlier

In [0]:
macSQLDF = spark.sql("""
SELECT *
FROM events
WHERE device = 'macOS'
ORDER By event_timestamp
""")

display(macSQLDF)

device ecommerce event_name event_previous_timestamp event_timestamp geo items traffic_source user_first_touch_timestamp user_id macOS List(null, null, null) mattresses null 1592539216262230 List(Waterbury, CT) List() direct 1592539216262230 UA000000103314644 macOS List(null, null, null) mattresses 1592322041626307 1592539218667144 List(Cedar Rapids, IA) List() email 1592321551374513 UA000000102690062 macOS List(null, null, null) reviews 1592538997090958 1592539218939615 List(Toledo, OH) List() google 1592538965425275 UA000000103314591 macOS List(null, null, null) guest 1592538390245546 1592539225913937 List(Los Angeles, CA) List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) facebook 1592535034918745 UA000000103313292 macOS List(null, null, null) main null 1592539226602157 List(Tulsa, OK) List() google 1592539226602157 UA000000103314648 macOS List(null, null, null) original 1592539226602157 1592539229480664 List(Tulsa, OK) List() google 1592539226602157 UA000000103314648 macOS List(null, null, null) main null 1592539231022266 List(New York, NY) List() google 1592539231022266 UA000000103314649 macOS List(null, null, null) guest 1592538809287537 1592539231776200 List(New York, NY) List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1), List(null, P_FOAM_K, King Foam Pillow, 79.0, 79.0, 1)) instagram 1592532057403935 UA000000103311782 macOS List(null, null, null) delivery 1592538781907651 1592539232284730 List(Rancho Cordova, CA) List() facebook 1592538781907651 UA000000103314549 macOS List(null, null, null) main null 1592539233151229 List(Phoenix, AZ) List() youtube 1592539233151229 UA000000103314650 macOS List(null, null, null) shipping_info 1592539225913937 1592539235059340 List(Los Angeles, CA) List(List(null, M_STAN_F, Standard Full Mattress, 945.0, 945.0, 1)) facebook 1592535034918745 UA000000103313292 macOS List(null, null, null) mattresses 1592232855039331 1592539236295422 List(Inkster, MI) List() email 1592232212733578 UA000000102435405 macOS List(null, null, null) add_item 1592539233151229 1592539237432046 List(Phoenix, AZ) List(List(null, M_STAN_Q, Standard Queen Mattress, 1045.0, 1045.0, 1)) youtube 1592539233151229 UA000000103314650 macOS List(null, null, null) main null 1592539289834596 List(Philadelphia, PA) List() google 1592539289834596 UA000000103314659 macOS List(null, null, null) add_item 1592538281060049 1592539295857466 List(Baltimore, MD) List(List(NEWBED10, P_FOAM_K, King Foam Pillow, 71.10000000000001, 79.0, 1)) email 1592496441998402 UA000000103170863 macOS List(null, null, null) main null 1592539299325832 List(Manhattan, KS) List() direct 1592539299325832 UA000000103314660 macOS List(null, null, null) main null 1592539302562339 List(Eveleth, MN) List() facebook 1592539302562339 UA000000103314661 macOS List(null, null, null) mattresses null 1592539305168803 List(Alliance, OH) List() google 1592539305168803 UA000000103314662 macOS List(null, null, null) guest 1592538188973932 1592539309500570 List(Delaware, OH) List(List(NEWBED10, M_STAN_K, Standard King Mattress, 1075.5, 1195.0, 1)) email 1592236739981372 UA000000102463131 macOS List(null, null, null) add_item 1592539236295422 1592539310482088 List(Inkster, MI) List(List(NEWBED10, M_STAN_T, Standard Twin Mattress, 535.5, 595.0, 1)) email 1592232212733578 UA000000102435405 macOS List(null, null, null) register 1592538695582992 1592539331750583 List(Reno, NV) List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) email 1592416477775320 UA000000102979935 macOS List(null, null, null) shipping_info 1592539331750583 1592539339785249 List(Reno, NV) List(List(NEWBED10, M_STAN_Q, Standard Queen Mattress, 940.5, 1045.0, 1)) email 1592416477775320 UA000000102979935 macOS List(null, null, null) mattresses null 1592539343571171 List(Miami, FL) List() instagram 1592539343571171 UA000000103314675 macOS List(null, null, null) register 1592538514988699 1592539346041388 List(Lansing, MI) List(List(null, M_STAN_K, Standard

-sandbox
%md ##### <img alt="Best Practice" title="Best Practice" style="vertical-align: text-bottom; position: relative; height:1.75em; top:0.3em" src="https://files.training.databricks.com/static/images/icon-blue-ribbon.svg"/> Check your work
- You should only see `macOS` values in the `device` column  
- The fifth row should be an event with timestamp `1592539226602157`

### Classroom Cleanup

In [0]:
%run ./Includes/Classroom-Cleanup


Dropped database and removed files in working directory